README
Linux-PipeMagic version 0.01
============================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2011 by Dave Lambley

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Linux::PipeMagic.

0.04   Thu 23 Jun 11:37:21 BST 2022
    - Update documentation.

0.04_01  Wed 22 Jun 14:29:02 BST 2022
    - Documentation fixes
    - Attempt to stub functions when not on Linux.

0.03  Tue Jan 31 11:55:54 2012
    - Add sendfile(2) support.

0.02  Tue Aug 9 08:46:00 2011
    - Add dependencies to MakeFile.PL.
    - Add further documentation.
    - Don't export symbols by default.

0.01  Tue Jul 26 20:41:33 2011
	- original version; created by h2xs 1.23 with options
		-n Linux::PipeMagic



Makefile.PL
use 5.010001;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Linux::PipeMagic',
    VERSION_FROM      => 'lib/Linux/PipeMagic.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    BUILD_REQUIRES    => {
        'Test::More'    => 0,
        'File::Slurp'   => 0,
        'File::Temp'    => 0,

    },
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Linux/PipeMagic.pm', # retrieve abstract from module
       AUTHOR         => 'Dave Lambley <dlambley@cpan.org>') : ()),
    LIBS              => [''], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
    META_MERGE => {
        resources => {
            repository => "http://github.com/davel/Linux-PipeMagic",
        },
        x_contributors => [
            "Simon Wilper <sxw\@chronowerks.de>",
        ],
    },
    LICENSE           => "perl",
    MIN_PERL_VERSION  => v5.10.1,
);
if  (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw( SPLICE_F_MOVE SPLICE_F_NONBLOCK SPLICE_F_MORE SPLICE_F_GIFT ));
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'Linux::PipeMagic',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}


lib/Linux/PipeMagic.pm
package Linux::PipeMagic;

use 5.010001;
use strict;
use warnings;
use Carp;

require Exporter;
use AutoLoader;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Linux::PipeMagic ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	systee
    syssplice
    syssendfile
    SPLICE_F_MOVE
    SPLICE_F_NONBLOCK
    SPLICE_F_MORE
    SPLICE_F_GIFT
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw();

our $VERSION = '0.04';

sub AUTOLOAD {
    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.

    my $constname;
    our $AUTOLOAD;
    ($constname = $AUTOLOAD) =~ s/.*:://;
    croak "&Linux::PipeMagic::constant not defined" if $constname eq 'constant';
    my ($error, $val) = constant($constname);
    if ($error) { croak $error; }
    {
	no strict 'refs';
	# Fixed between 5.005_53 and 5.005_61
#XXX	if ($] >= 5.00561) {
#XXX	    *$AUTOLOAD = sub () { $val };
#XXX	}
#XXX	else {
	    *$AUTOLOAD = sub { $val };
#XXX	}
    }
    goto &$AUTOLOAD;
}

require XSLoader;
XSLoader::load('Linux::PipeMagic', $VERSION);

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__

=head1 NAME

Linux::PipeMagic - Perl extension to use the zero copy IO syscalls

=head1 SYNOPSIS

  use Linux::PipeMagic qw/ systee syssplice syssendfile /;
  systee($fh_in, $fh_out, $num_bytes, 0);
  syssplice($fh_in, $fh_out, $num_bytes, 0);
  syssendfile($fh_out, $fh_in, $num_bytes);

=head1 DESCRIPTION

Linux::PipeMagic is a Perl XS wrapper around the L<splice(2)>, L<tee(2)> and L<sendfile(2)>
syscalls.  You can use them to efficiently copy data from one file descriptor to
another inside the kernel (splice), or to efficiently copy data from one pipe
to another (tee).

=head1 FUNCTIONS

=over

=item sysplice($fh_in, $fh_out, $num_bytes, $flags)

Copies C<$num_bytes> from C<$fh_in> to C<$fh_out>.  This is roughly equivalent to,

    sysread($fh_in, my $buf, $num_bytes);
    syswrite($fh_out, $buf);

although the transfer takes place entirely in kernel space.

Returns the number of bytes transferred.

=item systee($fh_in, $fh_out, $num_bytes, $flags)

Copies C<$num_bytes> from C<$fh_in> to C<$fh_out>.  The filehandles must both
be of type pipe.  This works similarly like C<syssplice> but does not advance the
read pointer in C<$fh_in>.

Returns the number of bytes transferred.

=item syssendfile($fh_out, $fh_in, $num_bytes)

Copies C<$num_bytes> from C<$fh_in> to C<$fh_out>.  With current versions of
Linux, C<$fh_in> must be a file opened for reading, and C<$fh_out> must be a
writable socket.  Note the different order of parameters compared to the other
functions.

=back

=head1 CONSTANTS

=over

=item *
SPLICE_F_MOVE

=item *
SPLICE_F_NONBLOCK

=item *
SPLICE_F_MORE

=item *
SPLICE_F_GIFT

=back

=head1 CAVEATS

Only Linux is supported, on other OSs the calls will fail with C<ENOSYS> and
the constants will not be available. L<tee(2)> and L<splice(2)> syscalls only
exist on Linux. L<sendfile(2)> is provided by the BSDs, however it takes different
parameters to the Linux call.

=head1 SEE ALSO

=over

=item *
L<http://github.com/davel/Linux-PipeMagic/>

=back

See the Linux manpages for more details on how splice and tee can be used,
including the flags.

=over

=item *
L<splice(2)>

=item *
L<tee(2)>

=item *
L<sendfile(2)>

=back

=head1 AUTHOR

Dave Lambley, E<lt>dlambley@cpan.org<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011, 2012, 2022 by Dave Lambley

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.

=cut
